crates/engine/src/prepare/trusted_documents.rs

```diff
@@ -115,13 +149,81 @@ impl<'ctx, R: Runtime> PrepareContext<'ctx, R> {
             }
-            (false, Some(ext), _) => {
-                if !self.engine.schema.settings.apq_enabled {
+            (TrustedDocumentsEnforcementMode::Allow, _, Some(doc_id)) => {
+                let Some(client_name) = client_name else {
+                    return Err(missing_client_name_error());
+                };
+
+                let query = request.query.as_deref();
+
+                Ok(ExtractedOperationDocument {
+                    // Reflect what actually gets executed. If there is an inline query document, it will always take priority.
+                    key: query
+                        .map({
+                            let operation_name = operation_name.clone();
+                            move |query| DocumentKey::Text {
+                                operation_name,
+                                document: Cow::Borrowed(query),
+                            }
+                        })
+                        .unwrap_or_else(|| DocumentKey::TrustedDocumentId {
+                            operation_name,
+                            client_name: Cow::Borrowed(client_name),
+                            doc_id: Cow::Borrowed(doc_id),
+                        }),
+                    content: DocumentOrFuture::Future(
+                        handle_trusted_document_query_and_check_that_query_matches(
+                            self.engine,
+                            client_name,
+                            Cow::Borrowed(doc_id),
+                            query,
+                        )
+                        .boxed(),
+                    ),
+                })
+            }
+            (TrustedDocumentsEnforcementMode::Allow, Some(ext), _) if !apq_enabled => {
+                let Some(client_name) = client_name else {
+                    return Err(missing_client_name_error());
+                };
+
+                let query = request.query.as_deref();
+                let doc_id: Cow<'_, str> = Cow::Owned(hex::encode(&ext.sha256_hash));
+
+                Ok(ExtractedOperationDocument {
+                    // Reflect what actually gets executed. If there is an inline query document, it will always take priority.
+                    key: query
+                        .map({
+                            let operation_name = operation_name.clone();
+                            |query| DocumentKey::Text {
+                                operation_name,
+                                document: Cow::Borrowed(query),
+                            }
+                        })
+                        .unwrap_or_else(|| DocumentKey::TrustedDocumentId {
+                            operation_name,
+                            client_name: Cow::Borrowed(client_name),
+                            doc_id: doc_id.clone(),
+                        }),
+                    content: DocumentOrFuture::Future(
+                        handle_trusted_document_query_and_check_that_query_matches(
```

Comment:
The check against the query shouldn't be hidden. The goal of the future is to prevent fetching a trusted document if we have the query in cache already. But here you're doing something different. If both the query and document id are provided, shouldn't you always check if they're consistent?

Reply:
or do we only want to check it if not in operation cache?